lib/std/tempfiles.nim

```diff
@@ -0,0 +1,138 @@
+#
+#
+#            Nim's Runtime Library
+#        (c) Copyright 2021 Nim contributors
+#
+#    See the file "copying.txt", included in this
+#    distribution, for details about the copyright.
+#
+
+## This module creates temporary files and directories.
+
+import std/[os, random]
+
+
+const
+  maxRetry = 10000
+  letters = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
+  nimTempPathLength {.intdefine.} = 8
+
+
+when defined(windows):
+  import std/winlean
+
+  var O_RDWR {.importc: "_O_RDWR", header: "<fcntl.h>".}: cint
+
+  proc c_fdopen(
+    filehandle: cint,
+    mode: cstring
+  ): File {.importc: "_fdopen",header: "<stdio.h>".}
+
+  proc open_osfhandle(osh: Handle, mode: cint): cint {.
+    importc: "_open_osfhandle", header: "<io.h>".}
+
+  proc close_osfandle(fd: cint): cint {.
+    importc: "_close", header: "<io.h>".}
+else:
+  import std/posix
+
+  proc c_fdopen(
+    filehandle: cint,
+    mode: cstring
+  ): File {.importc: "fdopen",header: "<stdio.h>".}
+
+
+proc safeOpen(filename: string): File =
+  when defined(windows):
+    let dwShareMode = FILE_SHARE_DELETE or FILE_SHARE_READ or FILE_SHARE_WRITE
+    let dwCreation = CREATE_NEW
+    let dwFlags = FILE_FLAG_BACKUP_SEMANTICS or FILE_ATTRIBUTE_NORMAL
+    let handle = createFileW(newWideCString(filename), GENERIC_READ or GENERIC_WRITE, dwShareMode,
+                              nil, dwCreation, dwFlags, Handle(0))
+
+    if handle == INVALID_HANDLE_VALUE:
+      raiseOSError(osLastError(), filename)
+
+    let fileHandle = open_osfhandle(handle, O_RDWR)
+    if fileHandle == -1:
+      discard closeHandle(handle)
+      raiseOSError(osLastError(), filename)
+
+    result = c_fdopen(fileHandle, "w+")
+    if result == nil:
+      discard close_osfandle(fileHandle)
+      raiseOSError(osLastError(), filename)
+  else:
+    let flags = posix.O_RDWR or posix.O_CREAT or posix.O_EXCL
+
+    let fileHandle = posix.open(filename, flags)
+    if fileHandle == -1:
+      raiseOSError(osLastError(), filename)
+
+    result = c_fdopen(fileHandle, "w+")
+    if result == nil:
+      discard posix.close(fileHandle)
+      raiseOSError(osLastError(), filename)
+
+template randomPathName(length: Natural): string =
+  var res = newString(length)
+  var state = initRand()
+  for i in 0 ..< length:
+    res[i] = letters[state.rand(61)]
+  res
+
+proc createTempFile*(prefix, suffix: string, dir = ""): tuple[fd: File, name: string] =
+  ## `createTempFile` creates a new temporary file in the directory `dir`. If `dir` is
+  ## the empty string, the default directory for temporary files
+  ## (`getTempDir <os.html#getTempDir>`_) will be used.
+  ## The temporary file name begins with `prefix` and ends with `suffix`.
+  ##
+  ## .. note:: It is the caller's responsibility to remove the file when no longer needed.
+  ##
+  ## `createTempFile` returns a file handle to an open file and the path of that file.
+  ## If failing to create a temporary file, `IOError` will be raised.
```

Comment:
Nit: consider formatting the doc comments for procs like this:

```
## Short description of what the proc does.
##
## Explanation of the details of the proc.
##
## Explanation of the procs exceptions.
##
## .. note:: ...

Reply:
fixed